blockchain/standalone/pow.go

```diff
@@ -218,3 +218,181 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
 	// The proof of work hash must be less than the target difficulty.
 	return checkProofOfWorkHash(powHash, target)
 }
+
+// CalcASERTDiff calculates an absolutely scheduled exponentially weighted
+// target difficulty for the given set of parameters using the algorithm defined
+// in DCP0011.
+//
+// The Absolutely Scheduled Exponentially weighted Rising Targets (ASERT)
+// algorithm defines an ideal schedule for block issuance and calculates the
+// difficulty based on how far the most recent block's timestamp is ahead or
+// behind that schedule.
+//
+// The target difficulty is set exponentially such that it is doubled or halved
+// for every multiple of the half life the most recent block is ahead or behind
+// the ideal schedule.
+//
+// The starting difficulty bits parameter is the initial target difficulty all
+// calculations use as a reference.  This value is defined on a per-chain basis.
+// It must be non-zero and less than or equal to the provided proof of work
+// limit or the function will panic.
+//
+// The time delta is the number of seconds that have elapsed between the most
+// recent block and an initial reference timestamp.
+//
+// The height delta is the number of blocks between the most recent block height
+// and an initial reference height.  It must be non-negative or the function
+// will panic.
+//
+// This function is safe for concurrent access.
+func CalcASERTDiff(startDiffBits uint32, powLimit *big.Int, targetSecsPerBlock,
+	timeDelta, heightDelta, halfLife int64) uint32 {
+
+	// Ensure parameter assumptions are not violated.
+	//
+	// 1. The starting target difficulty must be in the range [1, powLimit]
+	// 2. The height to calculate the difficulty for must come after the height
+	//    of the reference block
+	startDiff := CompactToBig(startDiffBits)
+	if startDiff.Sign() <= 0 || startDiff.Cmp(powLimit) > 0 {
+		panicf("starting difficulty %064x is not in the valid range [1, %064x]",
+			startDiff, powLimit)
+	}
+	if heightDelta < 0 {
+		panicf("provided height delta %d is negative", heightDelta)
+	}
+
+	// Calculate the target difficulty by multiplying the provided starting
+	// target difficulty by an exponential scaling factor that is determined
+	// based on how far ahead or behind the ideal schedule the given time delta
+	// is along with a half life that acts as a smoothing factor.
+	//
+	// Per DCP0011, the goal equation is:
+	//
+	//   nextDiff = min(max(startDiff * 2^((Δt - Δh*Ib)/halfLife), 1), powLimit)
```

Comment:
Yes, the change to per-block difficulty is very intentional.  I'm glad to see you that. The DCP (which isn't up yet) calls this out more explicitly (though it still needs a bit more refinement) along with the motivation behind it.

In short though, your observation is correct. It improves the responsiveness and allows the algorithm to better maintain the target block time (aka ideal block schedule). Another important aspect is that it means the network can more easily adjust to large swings in the hash power, notable large drops, versus the current EMA approach with a larger retarget interval since it only needs a single block at the no-longer-ideal difficulty versus potentially an entire interval.